File: BackEnd/src/domain/support-program.ts
import {
  GradeId,
  SkillId,
  SubjectId,
  SupportItemId,
  SupportProgramId,
  SupportProgramMilestoneId,
} from './ids';

export class SupportProgramItem {
  supportItemId: SupportItemId;
  orderInMilestone: number;
  name: string;
  type: string;
  skillsSupported: SkillId[];
  estimatedTime: number;
  difficulty: number;
}

export class SupportProgramMilestone {
  milestoneId: SupportProgramMilestoneId;
  milestoneNo: number;
  name: string;
  goal: string;
  requiredExerciseCount?: number;
  requiredQuizScore?: number;
  requiredQuizCount?: number;
  requiredAverageScore?: number;
  items: SupportProgramItem[];
}

export class SupportProgram {
  supportProgramId: SupportProgramId;
  programName: string;
  gradeId: GradeId;
  subjectId: SubjectId;
  targetSkill: SkillId;
  sourceUnitNo: number;
  sourceLessonNo: number;
  milestones: SupportProgramMilestone[];
}

File: Frontend/GP/Models/SupportProgram.cs
using System.ComponentModel.DataAnnotations;

namespace GP.Models
{
    public class SupportProgram : BaseEntity
    {
        [Key]

        public int SupportProgramId { get; set; }
        public string ProgramName { get; set; }
        public int GradeId { get; set; }
        public Grade Grade { get; set; }
        public int SubjectId { get; set; }
        public Subject Subject { get; set; }
        public string TargetSkill { get; set; }
        public int SourceUnitNo { get; set; }
        public int SourceLessonNo { get; set; }
        public List<SupportProgramMilestone> Milestones { get; set; }
    }
}
